src/naive_bayes/mod.rs

```diff
@@ -0,0 +1,239 @@
+    /// * `x` - data of shape NxM where N is number of data points to estimate and M is number of features.
+    /// Returns a vector of size N with class estimates.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let y_classes = self.distribution.classes();
+        let (rows, _) = x.shape();
+        let predictions = (0..rows)
+            .map(|row_index| {
+                let row = x.get_row(row_index);
+                let (prediction, _probability) = y_classes
+                    .iter()
+                    .map(|class| {
+                        (
+                            class,
+                            self.distribution.conditional_probability(*class, &row)
+                                * self.distribution.prior(*class),
+                        )
+                    })
+                    .max_by(|(_, p1), (_, p2)| p1.partial_cmp(p2).unwrap())
+                    .unwrap();
+                *prediction
+            })
+            .collect::<Vec<T>>();
+        let mut y_hat = M::RowVector::zeros(rows);
+        for (i, prediction) in predictions.iter().enumerate().take(rows) {
+            y_hat.set(i, *prediction);
+        }
+        Ok(y_hat)
+    }
+}
+
+/// Naive Bayes classifier for categorical features
+pub struct CategoricalNB<T: RealNumber> {
+    class_labels: Vec<T>,
+    class_probabilities: Vec<T>,
+    coef: Vec<Vec<Vec<T>>>,
+    feature_categories: Vec<Vec<T>>,
+}
+
+impl<T: RealNumber, M: Matrix<T>> NBDistribution<T, M> for CategoricalNB<T> {
+    fn prior(&self, k: T) -> T {
+        match self.class_labels.iter().position(|&t| t == k) {
+            Some(idx) => self.class_probabilities[idx],
+            None => T::zero(),
+        }
+    }
+
+    fn conditional_probability(&self, k: T, j: &M::RowVector) -> T {
+        match self.class_labels.iter().position(|&t| t == k) {
+            Some(idx) => {
+                let mut prob = T::one();
+                for feature in 0..j.len() {
+                    let value = j.get(feature);
+                    match self.feature_categories[feature]
+                        .iter()
+                        .position(|&t| t == value)
+                    {
+                        Some(_i) => prob *= self.coef[idx][feature][_i],
+                        None => return T::zero(),
+                    }
+                }
+                prob
+            }
+            None => T::zero(),
+        }
+    }
+
+    fn classes(&self) -> &Vec<T> {
+        &self.class_labels
+    }
+}
+
+impl<T: RealNumber> CategoricalNB<T> {
+    /// Fits the distribution to a NxM matrix where N is number of samples and M is number of features.
+    /// * `x` - training data.
+    /// * `y` - vector with target values (classes) of length N.
+    /// * `alpha` - Additive (Laplace/Lidstone) smoothing parameter (0 for no smoothing).
+    pub fn fit<M: Matrix<T>>(x: &M, y: &M::RowVector, alpha: T) -> Result<Self, Failed> {
```

Comment:
Let's make sure that alpha is an optional parameter. By default it should be 1.0.

Reply:
done!